### PR TITLE
Adding multi-cluster support documentation

### PR DESCRIPTION
**Pull request by subhankarc**
* Adding sections for Multi-Cluster support.

For better readability, https://github.com/cloudfoundry-incubator/service-fabrik-broker/blob/subhankarc-patch-gh-pages/inter-operator/architecture/basic.md

### CODE REVIEW

**Comment by subhankarc** [comment]
@amshuman-kr can you please have a look at this PR ?

File: inter-operator/architecture/basic.md [hunk 1]
```diff
@@ -531,6 +545,73 @@ status:
...
+  secretRef: 1-kubeconfig
+```
+where the secretRef looks like the following
+
+```yaml
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: vivek2-kubeconfig
```
Comment (vivekzhere) [inline inter-operator/architecture/basic.md]: name should be "1-kubeconfig" as the secretRef has that name

File: inter-operator/architecture/basic.md [hunk 2]
```diff
@@ -531,6 +545,73 @@ status:
...
+  namespace: interoperator
+data:
+  kubeconfig: <REDACTED_KUBECONFIG>
+```
+## Components within Interoperator
+Below, we discuss about the components of Service Fabrik Interoperator. Some components like the broker and the provisioner were already intruduced earlier. With Multi-Cluster deploy support, we bring in two new components, `MultiClusterDeployer` and `Scheduler` which are also described below.
```
Comment (vivekzhere) [inline inter-operator/architecture/basic.md]: introduced (spelling)

File: inter-operator/architecture/basic.md (hunk 1 above)
Comment (subhankarc) [reply inter-operator/architecture/basic.md]: done

File: inter-operator/architecture/basic.md [hunk 3]
```diff
@@ -531,6 +545,73 @@ status:
...
+Scalability is the main reason why one should use Multi-Cluster support. It gives you an option to add new clusters into your set of clusters and scale horizontally. There is always a limitation with the number of resources you can spawn in a cluster, even with maximum number of workers attached to a cluster, one can only scale to a finite number of worker node, which can hold a finite number of services(resources) only. Even with multiple worker nodes, APIServer(ETCD) also can become the bollteneck for scaling. Hence, for a production sceranio, multi-cluster support will be required so that services can be scheduled and spread across multiple clusters and can be scaled horizontally. Regarding the type of scheduling algorithms which are supported, we currently support round-robin and least-filled scheduler. We also plan to implement other schedulers which can be used. Schedulers are discussed later in the [schedulers](#schedulers) section.
+## New Custom Resources Introduced
+Along with the new custom resources like `SFService`, `SFPlan`, `SFServiceInstance` and `SFServiceBinding` which are discussed earlier, we also introduce `SFCluster` as a new CRD.
```
Comment (amshuman-kr) [inline inter-operator/architecture/basic.md]:
```suggestion
Along with the custom resources like `SFService`, `SFPlan`, `SFServiceInstance` and `SFServiceBinding` which are discussed earlier, we also introduce `SFCluster` as a new CRD.
```

File: inter-operator/architecture/basic.md (hunk 2 above)
Comment (subhankarc) [reply inter-operator/architecture/basic.md]: done

File: inter-operator/architecture/basic.md [hunk 4]
```diff
@@ -531,6 +545,73 @@ status:
...
+### SFCluster
+`SFCluster` is the CRD which stores the details of the cluster where provisioning of the service is needed. Depending on how many clusters need to be onboarded, we need to add resources for those. The structure of a sample resource look like the following.
```
Comment (amshuman-kr) [inline inter-operator/architecture/basic.md]:
```suggestion
`SFCluster` is the CRD which stores the details of the cluster where service instances are to be provisioned. One `SFCluster` CRD instance must be maintained for each cluster that is onboarded for provisioning service instances. The structure of a sample resource look like the following.
```

File: inter-operator/architecture/basic.md [hunk 5]
```diff
@@ -531,6 +545,73 @@ status:
...
+  namespace: interoperator
+data:
+  kubeconfig: <REDACTED_KUBECONFIG>
+```
+## Components within Interoperator
+Below, we discuss about the components of Service Fabrik Interoperator. Some components like the broker and the provisioner were already intruduced earlier. With Multi-Cluster deploy support, we bring in two new components, `MultiClusterDeployer` and `Scheduler` which are also described below.
+### Broker
+Broker was already introduced earlier, please read about it in the earlier section [here](#service-fabrik-inter-operator-broker)
+### MultiClusterDeployer
+This component is a set of [custom controller](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#custom-controllers). Below are the list of controllers it comprises of.
```
Comment (amshuman-kr) [inline inter-operator/architecture/basic.md]:
```suggestion
This component is a set of [custom controllers](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#custom-controllers). Below are the list of controllers it comprises of.
```

File: inter-operator/architecture/basic.md [hunk 6]
```diff
@@ -531,6 +545,73 @@ status:
...
+#### Provisioner Controller
+Provisioner Controller is the custom controller which watches on the `SFCluster` of the master cluster and deploys the [Provisioner](#provisioner) component in those clusters.
+#### Service Replicator
+Service Replicator is the custom controller which watches on the `SFClusters`, `SFServices` and `SFPlans` of the master cluster and copies the `SFServices` and `SFPlans` from master cluster to sister clusters.
+#### Service Instance Reconciler
+Service Instance Reconciler is the custom controller which watches across multiple clusters, part of the cluster registry(set of `SFClusters`) and reconciles a `SFServiceInstance` between master cluster and its assigned cluster, assigned by [Scheduler](#schedulers).
```
Comment (amshuman-kr) [inline inter-operator/architecture/basic.md]:
```suggestion
Service Instance Reconciler is the custom controller which watches across multiple clusters that are part of the cluster registry(set of `SFClusters`) and reconciles a `SFServiceInstance` between master cluster and its assigned cluster, assigned by [Scheduler](#schedulers).
```

File: inter-operator/architecture/basic.md [hunk 7]
```diff
@@ -531,6 +545,73 @@ status:
...
+#### Service Binding Reconciler
+Service Binding Reconciler is the custom controller which watches across multiple clusters, part of the cluster registry(set of `SFClusters`) and reconciles a `SFServiceBinding` between master cluster and its assigned cluster, assigned by [Scheduler](#schedulers).
+### Schedulers
+Schedulers are basically custom controller running on master cluster watching on `SFServiceInstances` and schedules/assigns them `clusterId` where the instance need to be provisioned, depending on the scheduling algorithm it implements. We currently have implemented the following set of schedulers described below. Activating a scheduler is config driven to be passed when someone deploys Inter-operator.
```
Comment (amshuman-kr) [inline inter-operator/architecture/basic.md]:
```suggestion
Schedulers are basically custom controller running on master cluster watching on `SFServiceInstances` and schedules/assigns them `clusterId` (the name of the corresponding `SFCluster` instance) of the cluster where the instance need to be provisioned, depending on the scheduling algorithm it implements. We currently have implemented the following set of schedulers described below. Activating a scheduler is config driven to be passed when someone deploys Inter-operator.
```

File: inter-operator/architecture/basic.md [hunk 8]
```diff
@@ -531,6 +545,73 @@ status:
...
+#### DefaultScheduler
+This is just a sample scheduler if one does not want multiple clusters in the setup. In that case, it schedules all the instances in the one cluster which is part of the setup.
```
Comment (amshuman-kr) [inline inter-operator/architecture/basic.md]:
```suggestion
This is just a sample scheduler suitable only for the single cluster setup. In that case, it schedules all the instances in the one cluster which is part of the setup. It is not suitable for the multi-cluster setup.
```

File: inter-operator/architecture/basic.md [hunk 9]
```diff
@@ -531,6 +545,73 @@ status:
...
+#### Round Robin Scheduler
+As the name suggests, round robin scheduler schedules instances in round robin fashion. At this point, it does not take care of capacity and if interoperator restarts, it starts scheduling from the beginning.
```
Comment (amshuman-kr) [inline inter-operator/architecture/basic.md]:
```suggestion
As the name suggests, round robin scheduler schedules instances in round robin fashion. At this point, it does not take care of capacity and if interoperator restarts, it loses the state about the next cluster to be scheduled and starts scheduling from the beginning. 
```

File: inter-operator/architecture/basic.md [hunk 10]
```diff
@@ -531,6 +545,73 @@ status:
...
+### Schedulers
+Schedulers are basically custom controller running on master cluster watching on `SFServiceInstances` and schedules/assigns them `clusterId` where the instance need to be provisioned, depending on the scheduling algorithm it implements. We currently have implemented the following set of schedulers described below. Activating a scheduler is config driven to be passed when someone deploys Inter-operator.
+#### DefaultScheduler
+This is just a sample scheduler if one does not want multiple clusters in the setup. In that case, it schedules all the instances in the one cluster which is part of the setup.
+#### Round Robin Scheduler
+As the name suggests, round robin scheduler schedules instances in round robin fashion. At this point, it does not take care of capacity and if interoperator restarts, it starts scheduling from the beginning. 
+#### Least filled Scheduler
+This scheduler schedules instance in the least filled cluster.
+### Provisioner
+Provisioner was also already introduced earlier, please read about it in the earlier section [here](#service-fabrik-inter-operator-provisioner)
```
Comment (amshuman-kr) [inline inter-operator/architecture/basic.md]: Should we add the comment that this is automatically deployed by the inter-operator?

File: inter-operator/architecture/basic.md [hunk 11]
```diff
@@ -531,6 +545,73 @@ status:
...
+## Deployment Flow
+Following are the flow for a deployment of Interoperator.
+1. When Interoperator is deployed initially, one deploys the [broker](#broker), [MultiClusterDeployer](#multiclusterdeployer) and the [Scheduler](#schedulers) component in a cluster, called as master cluster.
+2. After this, the operator should create the `SFServices`, `SFPlans` and `SFClusters` in the master cluster. `SFClusters` is simply the list/registry of all clusters where you want to provision the instances. We also refer to them as sister cluster interchangebly. Master cluster can also be part of the cluster registry and be a sister cluster in itself, if someone wants to use it for service provisioning as well.
+3. [Provisioner Controller](#provisioner-controller) then takes care of replicating the provisioner component to all sister clusters and [Service Replicator](#service-replicator) takes care of replicating the SFServices and SFPlans in all the clusters.
+
+Now the setup is ready for taking requests. We depict this in the picture below.
+![Inter-operator Deployment Flow](images/Deployment%20Flow.png)
```
Comment (amshuman-kr) [inline inter-operator/architecture/basic.md]: Can we visualize controllers and CRDs differently?

File: inter-operator/architecture/basic.md [hunk 12]
```diff
@@ -531,6 +545,73 @@ status:
...
+## Runtime Flow
+After the interoperator is ready and setup across multiple clusters as described [above](#deployment-flow), service instance and service binding can be created. When in the master cluster, broker creates an `SFServiceInstance`, Scheduler picks it up first and schedules/assigns a cluster where service needs to be provisioned. Then [Service Instance Reconciler](#service-instance-reconciler) reconciles that `SFServiceInstance` in the sister cluster where it is scheduled. Once that is done, [provisioner](#provisioner) residing in the sister cluster takes over and from then onwards, the process described in [service provisioning](#service-fabrik-inter-operator-provisioner-1) is followed. For another `SFServiceInstance`, it is again scheduled in one of the sister cluster and provisioner provisions the service there. The picture below describes the steps.
+![Inter-operator Runtime Flow](images/Runtime%20Flow.png)
```
Comment (amshuman-kr) [inline inter-operator/architecture/basic.md]: Can we highlight the difference from one frame to another left-to-right? The changes don't jump out.

File: inter-operator/architecture/basic.md (hunk 10 above)
Comment (subhankarc) [reply inter-operator/architecture/basic.md]: Added a line about it.

File: inter-operator/architecture/basic.md (hunk 11 above)
Comment (subhankarc) [reply inter-operator/architecture/basic.md]: With different color coding ? Or shall I put them inside APIServer shown as a state ?

File: inter-operator/architecture/basic.md (hunk 12 above)
Comment (subhankarc) [reply inter-operator/architecture/basic.md]: yeah make sense. I'll try to bring that in with an updated image.

File: inter-operator/architecture/basic.md (hunk 11 above)
Comment (subhankarc) [reply inter-operator/architecture/basic.md]: Made some changes in the diagrams. WDYTN?

File: inter-operator/architecture/basic.md (hunk 12 above)
Comment (subhankarc) [reply inter-operator/architecture/basic.md]: Made some changes in the diagrams. WDYTN?

File: inter-operator/architecture/basic.md (hunk 11 above)
Comment (amshuman-kr) [reply inter-operator/architecture/basic.md]: Color-coding as well as different shape? Or as you mentioned, inside the APIServer as a state.

Comment (subhankarc) [reply inter-operator/architecture/basic.md]: I changed the color coding and also kept them outside the interoperator box to denote that they are not a sub-component. Let me know if it looks good now.

**Pull request merged**